### ISSUE TITLE
Add Mask2Former to SMP

### ISSUE THREAD

**Issue by caxel-ap** [qubvel-org/segmentation_models.pytorch#1044]
Mask2Former model was introduced in the paper Masked-attention [Mask Transformer for Universal Image Segmentation](https://arxiv.org/abs/2112.01527) and first released in [this repository](https://github.com/facebookresearch/Mask2Former/).

Mask2Former addresses instance, semantic and panoptic segmentation with the same paradigm: by predicting a set of masks and corresponding labels. Hence, all 3 tasks are treated as if they were instance segmentation. Mask2Former outperforms the previous SOTA, [MaskFormer](https://arxiv.org/abs/2107.06278) both in terms of performance an efficiency by (i) replacing the pixel decoder with a more advanced multi-scale deformable attention Transformer, (ii) adopting a Transformer decoder with masked attention to boost performance without without introducing additional computation and (iii) improving training efficiency by calculating the loss on subsampled points instead of whole masks.

Papers with Code
https://paperswithcode.com/paper/masked-attention-mask-transformer-for

Paper:
https://arxiv.org/abs/2112.01527

HF Reference implementation:
https://huggingface.co/docs/transformers/main/en/model_doc/mask2former
https://github.com/huggingface/transformers/blob/main/src/transformers/models/mask2former/modeling_mask2former.py

**Comment by qubvel** [comment]
Thanks for opening an issue @caxel-ap 🤗  It might be a first instance segmentation model in the library, let's see if anyone would be a eager to contribute, I suppose it will be super impactful 👍

**Comment by caxel-ap** [comment]
Even just for semantic segmentation would be great to have in here someday, I’ve had good results using it in transformers for semantic with https://huggingface.co/facebook/mask2former-swin-large-ade-semantic

**Comment by ariG23498** [comment]
Hey @qubvel !

I would like to work on it. Is there a guideline on how to contribute to this repo? My todos would be:
1. Read the resources shared
2. Read contribution guideline if available
3. Start a draft PR and iterate on it.

Thanks!

**Comment by qubvel** [comment]
Hey @ariG23498! That's super cool, thanks for your interest 🤗 

At the moment there are no guidelines, unfortunately, but you can get inspiration from any of the existing models. The code for existing models is relatively small, so you can just copy `decoders/unet` and start from that point.

- There is no need to implement an encoder, as far as I understand Swin should be compatible with `timm` model.
- As suggested above we can start with **semantic segmentation** decoder and see if we can extend it to instance/panoptic as well. 
- The main idea is to have `decoder.py` and `model.py` files under `decoders/mask2former`. 

Just let me know what questions you will face and I will try to answer, and then add it to the docs 🤗

**Comment by Harsh-2005d** [comment]
@qubvel  Seeing that this has been inactive for some time, I'd like to work on it if it's still on the roadmap. Following the earlier discussion, I'd start by drafting a PR if you approve.
